SLC2, BSP alignment and POF version 2118 support

Shivansps opened the pull request:
Changed SLDC system to SLC2, this replaced the char at the beginning of the node for an int in order for the tree to be aligned.
Added BSP_Defpoints data alignment and SLDC to SLC2 conversion for pof version < 2118.

Shivansps commented:
Re-uploaded #2200

JohnAFernandez commented:
So that maybe we can finally get this done, it looks like travis just doesn't like this: ```
`/home/travis/build/scp-fs2open/fs2open.github.com/code/model/modelsinc.h:20:2: error: This should only be used internally by the model library.  See John if you think you need to include this elsewhere. [clang-diagnostic-error]
#error This should only be used internally by the model library.  See John if you think you need to include this elsewhere.`
```

Meaning that we probably just need to enable that to be used elsewhere, if I'm reading that correctly.

Shivansps commented:
Yeah the problem with that is that error makes no sence, that should only happen if, lets say, i edit dds.cpp and i include modelsync.h, etc

I added no includes, and all the code i added is inside the "modellib", the problem to me is modelsync.h its either being included before the modellib definition for some weird reason, or it is wrongly included somewhere and any change to it triggers the error.

Ill look into it a little more today to see if i can discover anything.

asarium commented:
The issue here is that clang-tidy tries to only parse that single header so the checked macro is not defined. You can remove that `#error` to fix the issue.

Shivansps commented:
Goober told me not to comment the error.

Goober5000 commented:
If the error only occurs for this specific PR because of clang, I think we can ignore it and merge the PR anyway.  It will not affect subsequent builds.

This means we can just go ahead and do a proper code review.

asarium commented on code/model/modelread.cpp:
```diff
@@ -1700,13 +1700,46 @@ int read_model_file(polymodel * pm, const char *filename, int n_subsystems, mode
+				{
+					pm->submodel[n].bsp_data_size = cfread_int(fp);
+					if (pm->submodel[n].bsp_data_size > 0) {
+						mprintf(("BSP_Data is being aligned.\n"));
```
This will log print will only spam the log file and is not really useful for mod debugging.

Shivansps replied:
Ok ill do that

asarium commented on code/model/modelread.cpp:
```diff
@@ -1738,14 +1771,37 @@ int read_model_file(polymodel * pm, const char *filename, int n_subsystems, mode
-				{
+				{   //ShivanSpS - if pof version is 2118 or higher ignore SLDC, otherwise convert it to slc2.
+					if (pm->version < 2118) {
+						mprintf(("SLDC data is being converted to SLC2.\n"));
```
Same here, please remove or convert to `nprintf`.

Shivansps replied:
ok

asarium commented on code/model/modelread.cpp:
```diff
@@ -1738,14 +1771,37 @@ int read_model_file(polymodel * pm, const char *filename, int n_subsystems, mode
+						pm->sldc_size = cfread_int(fp);
+						ubyte* sldc_tree, * slc2_tree;
+						sldc_tree = (ubyte*)vm_malloc(pm->sldc_size);
+						slc2_tree = (ubyte*)vm_malloc(pm->sldc_size * 2);
```
Please use `std::unique_ptr<ubyte[]>` for automatic deallocation.

Shivansps replied:
Ok im new to smart pointers, but you can do that when "cfread(sldc_tree, 1, pm->sldc_size, fp);" expects a void* on sldc_tree?

asarium replied:
You can still get the "raw" pointer via `.get()` if necessary.

asarium commented on code/model/modelread.cpp:
```diff
@@ -1738,14 +1771,37 @@ int read_model_file(polymodel * pm, const char *filename, int n_subsystems, mode
+						vm_free(sldc_tree);
+						vm_free(slc2_tree);
+					}
+				}
```
Maybe throw a warning in the else case since the presence of this chunk is not expected or valid in that case.

Shivansps replied:
The SLDC chunk needs to still be present on pof version 2118 or higher in order to provide backguard compatibility with older versions of FSO. It would just create a lot of unnecessary warnings.

asarium replied:
Ah, didn't know this was how that was supposed to work but that makes sense.

asarium commented on code/model/modelread.cpp:
```diff
@@ -5431,14 +5487,16 @@ void swap_bsp_data( polymodel * pm, void * model_ptr )
 void swap_sldc_data(ubyte * buffer)
 {
+//ShivanSpS - Changed type char for a type int for SLC2
 #if BYTE_ORDER == BIG_ENDIAN
```
Wait, we still support Big endian? Do we actually need that for ARM? I thought that could also handle little endian. If that is the case, maybe just remove this.

Shivansps replied:
Well, the function of swaping SLDC data is already in the code, so im forced to adjust it to the new node format, otherwise the game will crash on a big endian system.

asarium replied:
Yes but the point is if the big endian case is even still necessary since I think all of our supported systems are little endian. That is not a question to answer here though so there is nothing to change here. It was just something I noticed.

asarium commented on code/model/modelsinc.h:
```diff
@@ -12,6 +12,9 @@
 class polymodel;
+struct vec3d;
+struct matrix;
+typedef std::uint8_t ubyte;
```
We already have this typedef in `pstypes.h` to please only use that instead of repeating it here.

Shivansps replied:
This is also part of the problem with this specific header, whiout that you are going to get errors about missing these types in CI. For the record these types were already in use by the header, as i only added a definition.

Shivansps replied:
Using pragma message this header now works, but it will fail again if i remove those forward declarations.

asarium reviewed (changes requested):
How extensively has this been testing with existing model data? This has the potential of breaking quite a lot of stuff so maybe create a test build and have some people do some smoke testing on existing mods to check if this works as intended.

asarium commented:
@Goober5000 Simply ignoring static code analysis is not a good way of using our CI results.
If the header cannot be compiled on its own, then it is, in my opinion, not a good header and the `#error` should be removed.

Shivansps commented:
As for testing, i did that a year when i wrote this, i created two missions, one that loaded every ship on FS2 with MediaVPS, and another one to test the shield collision was still working properly, i recreated them because i lost them.
Ive uploaded them along with a test build https://www.dropbox.com/s/0utr2rsgscsb916/RP%202691%20Test.rar?dl=0

Goober5000 commented:
> @Goober5000 Simply ignoring static code analysis is not a good way of using our CI results.

This isn't ignoring it.  It's carefully considering it and knowing when to follow it (which is most of the time) and when it is giving a false positive.

What if we changed it to `#pragma message`?

asarium commented:
Yes, I think a pragma would be the better solution here.

JohnAFernandez commented:
Hey this got merged by accident when we merged the compressor!
